### ISSUE TITLE
🐛 Running off of Docker gives failure for node-fetch

### ISSUE THREAD

**Issue by cshadd** [realcoloride/node_characterai#4]
Hi again,

I am trying to run this through Docker and I am getting  this error:

![image](https://user-images.githubusercontent.com/11986547/220655592-b4d80158-41f7-4440-891b-66708897f97a.png)

I referenced this in https://github.com/realcoloride/node_characterai/issues/2#issuecomment-1439100047

I assume node-fetch is somehow not being installed in Docker.

Thanks for your hard work.

**Comment by realcoloride** [comment]
Hello, have you tried doing
`npm install node-fetch`

```
import fetch from "node-fetch";
globalThis.fetch = fetch
```

In the global scope?

https://stackoverflow.com/questions/48433783/referenceerror-fetch-is-not-defined
There's other possible solutions here. Could you try them out and tell me which one worked for you?

**Comment by eliangerard** [comment]
Same problem again, and the ideal solution would be to install node 18. However, my problem was that due to library issues, it can't be installed on Amazon Linux 2. So the solution would be to install node 17 and add the flag --experimental-fetch to make it work (again, only if you can't install Node 18).

If you follow what @realcoloride said, you'll need to restructure your project because the new versions of fetch don't work with CommonJS, which means you can't use const foo = require('foo'). You can install version 2 of node-fetch with npm install node-fetch@2, which supports CommonJS, and then use the following code:

```
const fetch = require('node-fetch');
globalThis.fetch = fetch;
```

However, when I did this, I got the error "Error: Failed to fetch a lazy token.", when I use node 18/17 with flag it works perfectly, with guest or token mode, so it could be the fetch version.

To the dear developer:

I suggest that, if you want to make the package to work with node versions older than 18 natively, you would need to add the import of fetch from your project, investigate why the "Error: Failed to fetch a lazy token" is appearing with older versions of fetch, use another package instead of fetch or just require node 18. 
Versions of node-fetch greater than 2 do not support CommonJS. I tried to add it myself from a fork, but I got the error I mentioned and I don't have much time to keep trying (And the knowledge). I wanted to try with the latest version of node-fetch 3.3.0 that comes with Node 18 and does work, but it can only be explicity added with import and not with require, and I don't know much about this topic of CommonJS and ES Modules.

In summary, for now, the package works ideally in Node.js v18. If you can't use that version, use version 17 and add --experimental-fetch, for example: **node --experimental-fetch index.js**
For versions older than 17, with the fix of globalThis.fetch = fetch; throws "Error: Failed to fetch a lazy token."

I apologize if my English is hard to understand. ChatGPT helped me write this. And this.

**Comment by realcoloride** [comment]
Hello!

Thank you for bringing this answer, it helps me a lot!

I understand the limitations you have and your context.

The reason why it fails to fetch lazy token might be because the UUIDv4
library is not imported correctly (it uses require)

If you have found a workaround at the moment, use it until we found a
viable solution.

I will keep you updated along with the latest bug issue.

Le jeu. 23 févr. 2023 à 00:02, Elian Gerard ***@***.***> a
écrit :

**Comment by cshadd** [comment]
I have migrated to Node version 18 for the project, and it works. There are limitations on fetch for projects less than Node 17 as you mentioned.

Everything is working but I still will probably experiment and find a workaround and send a PR if I can get one working so we can have backwards compatibility with 17 and 16.

**Comment by realcoloride** [comment]
All good!

If you find a workaround it can be interesting to share it for those in help.
Sadly, I do not use older Node versions, but in the meantime, I do believe that the reason why its not working on older versions on docker may be because the uuid library is not imported library.

Try to do the same steps you've done for `node-fetch` with `uuid` and see how it runs.

Have a good day!
